x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointsShardAction.java

```diff
@@ -0,0 +1,242 @@
+            return waitForAdvance;
+        }
+
+        public long currentCheckpoint() {
+            return currentCheckpoint;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            shardId.writeTo(out);
+            out.writeBoolean(waitForAdvance);
+            out.writeLong(currentCheckpoint);
+            out.writeTimeValue(pollTimeout);
+        }
+    }
+
+    public static class TransportGetGlobalCheckpointAction extends TransportSingleShardAction<Request, Response> {
+
+        private final IndicesService indicesService;
+
+        @Inject
+        public TransportGetGlobalCheckpointAction(
+            ThreadPool threadPool,
+            ClusterService clusterService,
+            TransportService transportService,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            IndicesService indicesService
+        ) {
+            super(
+                NAME,
+                threadPool,
+                clusterService,
+                transportService,
+                actionFilters,
+                indexNameExpressionResolver,
+                Request::new,
+                ThreadPool.Names.GENERIC
+            );
+            this.indicesService = indicesService;
+        }
+
+        @Override
+        protected Response shardOperation(Request request, ShardId shardId) {
+            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+            final SeqNoStats seqNoStats = indexShard.seqNoStats();
+            return new Response(seqNoStats.getGlobalCheckpoint());
+        }
+
+        @Override
+        protected void asyncShardOperation(Request request, ShardId shardId, ActionListener<Response> listener) throws IOException {
+            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+            final SeqNoStats seqNoStats = indexShard.seqNoStats();
+
+            if (request.waitForAdvance() && request.currentCheckpoint() >= seqNoStats.getGlobalCheckpoint()) {
+                indexShard.addGlobalCheckpointListener(
+                    request.currentCheckpoint() + 1,
+                    new GlobalCheckpointListeners.GlobalCheckpointListener() {
+
+                        @Override
+                        public Executor executor() {
+                            return threadPool.executor(ThreadPool.Names.GENERIC);
+                        }
+
+                        @Override
+                        public void accept(final long g, final Exception e) {
+                            if (g != UNASSIGNED_SEQ_NO) {
+                                assert request.currentCheckpoint() <= g : shardId
```

Comment:
I think we expect a strictly higher number only?

```suggestion
                                assert request.currentCheckpoint() < g : shardId
```